mm2src/mm2_metrics/src/mm_metrics.rs

```diff
@@ -0,0 +1,575 @@
+}
+
+pub struct TagObserver;
+
+impl TagObserver {
+    /// This is for collecting and logging `prepare_tag_metrics` to dashboard.
+    ///
+    /// Used with `init_with_dashboard`.
+    pub async fn log_tag_metrics(log_state: LogWeak, recorder: Arc<MmRecorder>, interval: f64) {
+        loop {
+            Timer::sleep(interval).await;
+            let log_state = match LogArc::from_weak(&log_state) {
+                Some(log_arc) => log_arc,
+                _ => {
+                    return;
+                },
+            };
+
+            log!(">>>>>>>>>> DEX metrics <<<<<<<<<");
+
+            Self::prepare_tag_metrics(&recorder)
+                .into_iter()
+                .for_each(|(labels, name_value_map)| {
+                    let tags = labels_to_tags(labels.iter());
+                    let message = name_value_map_to_message(&name_value_map);
+                    log_state.log_deref_tags("", tags, &message);
+                });
+        }
+    }
+
+    /// Prepare tag metrics for logging in `log_tag_metrics`.
+    fn prepare_tag_metrics(recorder: &MmRecorder) -> HashMap<MetricLabels, MetricNameValueMap> {
+        let mut output = HashMap::new();
+
+        for (key, counter) in recorder.registry.get_counter_handles() {
+            let value = counter.get_inner().load(Ordering::Acquire);
+            map_metrics_to_prepare_tag_metric_output(key, PreparedMetric::Unsigned(value), &mut output);
+        }
+
+        for (key, gauge) in recorder.registry.get_gauge_handles() {
+            let value = f64::from_bits(gauge.get_inner().load(Ordering::Acquire));
+            map_metrics_to_prepare_tag_metric_output(key, PreparedMetric::Float(value), &mut output);
+        }
+
+        for (key, histo) in recorder.registry.get_histogram_handles() {
+            if let Some(values) = MmHistogram::new(&histo.get_inner().data()) {
+                map_metrics_to_prepare_tag_metric_output(key, PreparedMetric::Histogram(values), &mut output);
+            }
+        }
+        output
+    }
+}
+
+/// Used for parsing metrics to `prepare_tag_metric_output`
+fn map_metrics_to_prepare_tag_metric_output(
+    key: Key,
+    value: PreparedMetric,
+    output: &mut HashMap<MetricLabels, MetricNameValueMap>,
+) {
+    let (metric_name, labels) = key.into_parts();
+    output
+        .entry(labels)
+        .or_insert_with(MetricNameValueMap::new)
+        .insert(metric_name.as_str().to_string(), value);
+}
+
+pub(crate) fn labels_to_tags(labels: Iter<Label>) -> Vec<Tag> {
+    labels
+        .map(|label| Tag {
+            key: label.clone().into_parts().0.to_string(),
+            val: Some(label.value().to_string()),
+        })
+        .collect()
+}
+
+/// Used for parsing `MetricNameValueMap` into Message(loggable string).
+pub(crate) fn name_value_map_to_message(name_value_map: &MetricNameValueMap) -> String {
+    name_value_map
+        .iter()
+        .sorted_by(|x, y| x.partial_cmp(y).expect("sorting faulted"))
```

Comment:
we could use `Less` instead since(if I'm correct) [f64::NAN](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html#tymethod.partial_cmp) seems to be the only possible data that can be passed In.

```rust
.sorted_by(|x, y| x.partial_cmp(y).unwrap_or(Ordering::Equal))
```

else I think it's better to use `None` @sergeyboyko0791

Reply:
Good idea
> else I think it's better to use None